src/resolve/resolved-dns-browse-services.c

```diff
@@ -0,0 +1,779 @@
+        int r;
+
+        assert(m);
+
+        if (!answer)
+                return 0;
+
+        HASHMAP_FOREACH(sb, m->dns_service_browsers) {
+
+                r = dns_answer_match_key(answer, sb->key, NULL);
+                if (r < 0)
+                        return log_error_errno(
+                                        r,
+                                        "Failed to match answer key with service browser's key: %m");
+                if (r == 0)
+                        continue;
+
+                r = mdns_browser_revisit_cache(sb, owner_family);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to revisit cache for service browser: %m");
+        }
+
+        return 0;
+}
+
+static void mdns_browse_service_query_complete(DnsQuery *q) {
+        _cleanup_(dns_service_browser_unrefp) DnsServiceBrowser *sb = NULL;
+        _cleanup_(dns_query_freep) DnsQuery *query = q;
+        int r;
+
+        assert(query);
+        assert(query->manager);
+
+        if (query->state != DNS_TRANSACTION_SUCCESS)
+                return;
+
+        sb = dns_service_browser_ref(query->service_browser_request);
+        if (!sb)
+                return;
+
+        r = dns_answer_match_key(query->answer, sb->key, NULL);
+        if (r < 0)
+                return (void) log_error_errno(r,
+                                "Failed to match answer key with service browser's key: %m");
+
+        if (r == 0)
+                return;
+
+        r = mdns_browser_revisit_cache(sb, query->answer_family);
+        if (r < 0)
+                return (void) log_error_errno(r, "Failed to revisit cache for service browser: %m");
+
+        /* When the query is answered from cache, we only get answers for one
+         * answer_family i.e. either ipv4 or ipv6. We need to perform another
+         * cache lookup for the other answer_family */
+        if (query->answer_query_flags == SD_RESOLVED_FROM_CACHE) {
+                r = mdns_browser_revisit_cache(sb, query->answer_family == AF_INET ? AF_INET6 : AF_INET);
+                if (r < 0) {
+                        log_error_errno(r, "Failed to revisit cache for service browser: %m");
+                        return;
+                }
+        }
+}
+
+static int mdns_next_query_schedule(sd_event_source *s, uint64_t usec, void *userdata) {
+        _cleanup_(dns_service_browser_unrefp) DnsServiceBrowser *sb = NULL;
+        _cleanup_(dns_query_freep) DnsQuery *q = NULL;
+        int r;
+
+        assert(userdata);
+        assert_se(sb = dns_service_browser_ref(userdata));
+
+        /* Enable the answer from the cache for the very first query */
+        if (sb->delay == 0)
+                SET_FLAG(sb->flags, SD_RESOLVED_NO_CACHE, false);
+
+        /* Set the flag indicating that the query is continuous.
+         * RFC 6762 Section 5.2 outlines timing requirements for continuous queries.
+         */
+        SET_FLAG(sb->flags, SD_RESOLVED_QUERY_CONTINUOUS, true);
```

Comment:
when we want to set or unset a flag to some fixed value we usually do not use SET_FLAG(). We only use that if we have a boolean variable based on which we want to set a flag...

hence, maybe just do

```c
sb->flags |= SD_RESOLVED_QUERY_CONTINOUS;
```

that's probably easier to read for most C developers.